CS refactor

coat-soup opened the pull request:
The octree is currently loaded completely at once, which is pretty slow and mostly unnecessary. Considering shifting it to loading and saving only the chunks needed for rendering when they are relevant. That way only relevant bits are ever computed.

<img width="1028" height="925" alt="image" src="https://github.com/user-attachments/assets/1d1ec588-269a-4b21-9932-e9eec0b3597d" />

Closes #17

coat-soup committed:
data octree creation

coat-soup committed:
moved all code to runtime

it kept crashing the editor lol

RaulPPelaez commented:
AFICT your tree construction is sequential and CPU, if it becomes a bottleneck you can always port that to the GPU, say with a compute shader. Not the easiest but feasible https://research.nvidia.com/sites/default/files/pubs/2012-06_Maximizing-Parallelism-in/karras2012hpg_paper.pdf

coat-soup commented:
I've been putting it off because I'm lazy and don't want to set up a gazillion buffers for all the cell data lol, but that is an excellent idea. Thanks for the resource!

coat-soup committed:
chunk rendering

coat-soup committed:
larger leaf node chunk loading

coat-soup committed:
barycentric interpolation

coat-soup committed:
chunk data saving and loading

coat-soup committed:
dynamic chunk loading

coat-soup committed:
dynamic octree & chunk lods

this is pretty much the entire system - just not threaded (so there's stuttering)

RaulPPelaez commented:
I installed godot-mono 4.5.1:
```shell
% brew install godot-mono
==> Auto-updating Homebrew...
Adjust how often this is run with `$HOMEBREW_AUTO_UPDATE_SECS` or disable with
`$HOMEBREW_NO_AUTO_UPDATE=1`. Hide these hints with `$HOMEBREW_NO_ENV_HINTS=1` (see `man brew`).
==> Auto-updated Homebrew!
Updated 2 taps (homebrew/core and homebrew/cask).
==> New Formulae
hayagriva: Bibliography management tool
klog: Command-line tool for time tracking in a human-readable, plain-text file format
pixlet: App runtime and UX toolkit for pixel-based apps
talm: Manage Talos Linux configurations the GitOps way
tronbyt-server: Manage your apps on your Tronbyt (flashed Tidbyt) completely locally

You have 1 outdated cask installed.

==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-cask/1b16464e3ca7dc1abee174d00de45a68251a4f34/Casks/g/godot-mono.r
############################################################################################################################### 100.0%
==> Downloading https://github.com/godotengine/godot/releases/download/4.5.1-stable/Godot_v4.5.1-stable_mono_macos.universal.zip
==> Downloading from https://release-assets.githubusercontent.com/github-production-release-asset/15634981/6827be97-a79e-4ac8-8719-0fe
############################################################################################################################### 100.0%
==> Installing dependencies: dotnet-sdk
==> Downloading https://builds.dotnet.microsoft.com/dotnet/Sdk/10.0.101/dotnet-sdk-10.0.101-osx-arm64.pkg
############################################################################################################################### 100.0%
==> Installing Cask dotnet-sdk
==> Running installer for dotnet-sdk with `sudo` (which may request your password)...
Password:
installer: Package name is Microsoft .NET SDK 10.0.101 (arm64)
installer: Installing at base path /
installer: The install was successful.
==> Linking Binary 'dotnet' to '/opt/homebrew/bin/dotnet'
🍺  dotnet-sdk was successfully installed!
==> Installing Cask godot-mono
==> Moving App 'Godot_mono.app' to '/Applications/Godot_mono.app'
==> Linking Binary 'godot-mono.wrapper.sh' to '/opt/homebrew/bin/godot-mono'
🍺  godot-mono was successfully installed!
```

Then opening the project yields some errors:
```shell
% godot-mono project.godot
Godot Engine v4.5.1.stable.mono.official.f62fdbde1 - https://godotengine.org
Metal 3.2 - Forward+ - Using Device #0: Apple - Apple M4 Pro (Apple9)

GdUnit4 TCP Server:	Successfully started checked port: 31002
SCRIPT ERROR: Invalid call. Nonexistent function 'IsApiLoaded' in base 'CSharpScript'.
          at: is_api_loaded (res://addons/gdUnit4/src/dotnet/GdUnit4CSharpApiLoader.gd:72)
          GDScript backtrace (most recent call first):
              [0] is_api_loaded (res://addons/gdUnit4/src/dotnet/GdUnit4CSharpApiLoader.gd:72)
              [1] _enter_tree (res://addons/gdUnit4/plugin.gd:33)
No GdUnit4Net found.
Loading GdUnit4 Plugin success
Recovery last test session ..
Recovery last test session successfully, 5 tests restored.
ERROR: Node not found: "../MeshGenerator" (relative to "Node").
   at: get_node (scene/main/node.cpp:1908)
   GDScript backtrace (most recent call first):
       [0] colour_mesh (res://planet-sim/visuals/cell_data_visualiser.gd:24)
       [1] @vis_type_setter (res://planet-sim/visuals/cell_data_visualiser.gd:9)
ERROR: Node not found: "../SimulationPipeline" (relative to "Node").
   at: get_node (scene/main/node.cpp:1908)
   GDScript backtrace (most recent call first):
       [0] colour_mesh (res://planet-sim/visuals/cell_data_visualiser.gd:25)
       [1] @vis_type_setter (res://planet-sim/visuals/cell_data_visualiser.gd:9)
SCRIPT ERROR: Invalid access to property or key 'polyhedron' on a base object of type 'null instance'.
          at: CellDataVisualiser.colour_mesh (res://planet-sim/visuals/cell_data_visualiser.gd:29)
          GDScript backtrace (most recent call first):
              [0] colour_mesh (res://planet-sim/visuals/cell_data_visualiser.gd:29)
              [1] @vis_type_setter (res://planet-sim/visuals/cell_data_visualiser.gd:9)
(0.0, 0.0, 0.1569, 0.0039)
generated 10242 faces
SCRIPT ERROR: Out of bounds get index '0' (on base: 'Array[CellData]')
          at: CellDataVisualiser.colour_mesh (res://planet-sim/visuals/cell_data_visualiser.gd:35)
          GDScript backtrace (most recent call first):
              [0] colour_mesh (res://planet-sim/visuals/cell_data_visualiser.gd:35)
              [1] _ready (res://planet-sim/visuals/cell_data_visualiser.gd:20)
SCRIPT ERROR: Parse Error: Static function "IsPointInSphericalTriangle()" not found in base "CSharpScript".
          at: GDScript::reload (res://test/test_terrain_chunk.gd:13)
SCRIPT ERROR: Parse Error: Static function "IsPointInSphericalTriangle()" not found in base "CSharpScript".
          at: GDScript::reload (res://test/test_terrain_chunk.gd:25)
SCRIPT ERROR: Parse Error: Static function "TriArea()" not found in base "CSharpScript".
          at: GDScript::reload (res://test/test_terrain_chunk.gd:34)
SCRIPT ERROR: Parse Error: Static function "TriArea()" not found in base "CSharpScript".
          at: GDScript::reload (res://test/test_terrain_chunk.gd:45)
SCRIPT ERROR: Parse Error: Static function "TriArea()" not found in base "CSharpScript".
          at: GDScript::reload (res://test/test_terrain_chunk.gd:46)
ERROR: Failed to load script "res://test/test_terrain_chunk.gd" with error "Parse error".
   at: load (modules/gdscript/gdscript.cpp:3041)
```

The project then looks ok:
<img width="1338" height="828" alt="imagen" src="https://github.com/user-attachments/assets/724f0cba-c10c-4542-a256-9dbbab8c0170" />

RaulPPelaez commented:
These errors seem to be reported in your CI, but the check is marked as successful anyway:
```shell

update_scene_groups: end
reimport: begin: (Re)Importing Assets steps: 20
	reimport: step 0: Executing post-reimport operations...
reimport: end
loading_editor_layout: begin: Loading editor steps: 5
	loading_editor_layout: step 0: Loading editor layout...
loading_editor_layout: end
ERROR: Node not found: "../MeshGenerator" (relative to "Node").
   at: get_node (scene/main/node.cpp:1877)
ERROR: Node not found: "../SimulationPipeline" (relative to "Node").
   at: get_node (scene/main/node.cpp:1877)
SCRIPT ERROR: Invalid access to property or key 'polyhedron' on a base object of type 'null instance'.
          at: CellDataVisualiser.colour_mesh (res://planet-sim/visuals/cell_data_visualiser.gd:29)
(0.0, 0.0, 0.1569, 0.0039)
generated 10242 faces
SCRIPT ERROR: Out of bounds get index '0' (on base: 'Array[CellData]')
          at: CellDataVisualiser.colour_mesh (res://planet-sim/visuals/cell_data_visualiser.gd:35)
```

coat-soup committed:
cell visualiser export variable fix

coat-soup commented:
> These errors seem to be reported in your CI, but the check is marked as successful anyway:

These were editor errors the unit testing doesn't interact with, so it passed them. They didn't actually do anything, but I have fixed it. The message should now be gone for the CI and the editor when loading the sim scene.


> I installed godot-mono 4.5.1:

Please note this project is using 4.**_4_**.1. A link can be found in the readme. Most of your errors are likely linked to this, as c# support is constantly being updated. If not, please remember to build the c# solution with this little hammer icon, because it doesn't happen automatically if just running editor scripts. I will add a note to the readme.

<img width="700" alt="buildcsicon" src="https://github.com/user-attachments/assets/0612b147-fae8-47e3-8c0e-c4e5850772b0" />

coat-soup committed:
separate chunk thread

coat-soup committed:
terrain material improvement

coat-soup committed:
idw testing &  height caches

coat-soup committed:
chunk generation multithreading

coat-soup committed:
player rotation fix & game stats ui

coat-soup commented:
That pretty much wraps up the refactor, and puts the generation system in a pretty solid spot. There are still some performance optimisations that can be looked at later down the line, but I'm really happy with the current state.

Showcase video: https://www.youtube.com/watch?v=UFV0Lnyb_A0

Next steps are trees/foliage, climate integration, perhaps some terrain shaping, and quite importantly surface nets (#15)
